src/Servers.jl

```diff
@@ -130,278 +109,196 @@ function check_rate_limit(tcp;
         else
             @warn "reuseaddr=true may not be supported on this platform: $(Sys.KERNEL)"
-            Sockets.bind(tcpserver, host.host, host.port; reuseaddr=true)
+            Sockets.bind(tcpserver, inet.host, inet.port; reuseaddr=true)
         end
         Sockets.listen(tcpserver)
     else
-        tcpserver = Sockets.listen(host)
-        tcpref[] = tcpserver
+        tcpserver = Sockets.listen(inet)
     end
+    verbose && @info "Listening on: $host:$port"
 
-    if host isa Sockets.InetAddr # build debugging info
-        hostname = string(host.host)
-        hostport = string(host.port)
-    else
-        hostname = string(host)
-        hostport = ""
+    if tcpisvalid === nothing
+        tcpisvalid = ratelimit === nothing ? x->true : x->check_rate_limit(x, ratelimit)
     end
-    listenloop(f, tcpserver, sslconfig, hostname, hostport, pipeline_limit, require_ssl_verification, tcpisvalid, connectioncounter; kw...)
+
+    return listenloop(handler, Server2(sslconfig, tcpserver, string(host), string(port)), tcpisvalid,
+        connectioncounter, reuse_limit, readtimeout, verbose)
 end
 
-function listenloop(f, tcpserver, sslconfig, hostname, hostport, pipeline_limit, require_ssl_verification, tcpisvalid, connectioncounter; kw...)
-    try
-        id = 0
-        while isopen(tcpserver)
-            try
-                io = accept(tcpserver)
-                if !tcpisvalid(io; kw...)
-                    @info "Accept-Reject:  $io"
-                    close(io)
-                    continue
-                end
-                io = getsslcontext(io, sslconfig)
-                let i=id, conn = Connection(hostname, hostport, pipeline_limit, 0, require_ssl_verification, io)
-                    @async try
-                        @info "Accept ($i):  $conn"
-                        connectioncounter[] += 1
-                        handle_connection(f, conn; kw...)
+"main server loop that accepts new tcp connections and spawns async threads to handle them"
+function listenloop(h, server,
+    tcpisvalid=x->true, connectioncounter=Ref(0),
+    reuse_limit::Int=1, readtimeout::Int=0, verbose::Bool=false)
+    count = 1
+    while isopen(server)
+        try
+            io = accept(server)
+            io === nothing && continue
+            if !tcpisvalid(io)
+                verbose && @info "Accept-Reject:  $io"
+                close(io)
+                continue
+            end
+            connectioncounter[] += 1
+            conn = Connection(server.hostname, server.hostport, 0, 0, true, io)
+            let io=io, count=count
+                @async begin
+                    try
+                        verbose && @info "Accept ($count):  $conn"
+                        handle(h, conn, reuse_limit, readtimeout)
+                        verbose && @info "Closed ($count):  $conn"
                     catch e
-                        @error "Error ($i):  $conn" exception=(e, stacktrace(catch_backtrace()))
```

Comment:
Was logging `$conn` not a useful way to see which connection caused the error?

Reply:
In the error case, it can be very difficult to track which connection had the error when the same client is making all the requests.

Reply:
Ok, but I'm not sure I follow why it helps to remove the `conn` from the error message ?
Sometimes, when there are many different clients it is good to see which one caused the error.
Also, the show function for `conn` has information about transaction counts read/write busy state etc that can be useful for debugging.